server-phoenix/test/helios_web/controllers/web_hooks/slack_controller_test.exs

```diff
@@ -105,6 +108,337 @@ defmodule HeliosWeb.WebHooks.SlackControllerTest do
+    body = %{
+      "channel" => "asdf"
+    }
+
+    request_body = URI.encode_query(body)
+
+    headers = [
+      {"Accept", "application/json"},
+      {"Authorization", "Bearer asdf"},
+      {"Content-Type", "application/x-www-form-urlencoded"}
+    ]
+
+    with_mock HTTPoison,
+      post!: fn "https://slack.com/api/conversations.info",
+                request_body,
+                headers,
+                follow_redirect: true ->
+        %HTTPoison.Response{
+          body: "{
+            \"ok\": \"true\",
+            \"channel\": {
+                \"name\": \"general\"
+            }
+        }"
+        }
+      end do
+      conn
+      |> put_req_header("content-type", "application/json")
+      |> post(
+        Routes.slack_path(conn, :handle),
+        %{
+          token: "Sim-123456",
+          team_id: "TXXXXXXXX",
+          api_app_id: "AXXXXXXXX",
+          event: %{
+            client_msg_id: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
+            type: "message",
+            text: "hi",
+            user: "Priscila",
+            ts: "155984192.001600",
+            channel: "asdf",
+            event_ts: "1559841925.001600",
+            channel_type: "channel"
+          },
+          type: "event_callback",
+          event_id: "<%= @event_id %>",
+          event_time: 1_559_841_923,
+          authed_users: ["UXXXXXXXX", "UXXXXXXXX"],
+          slack: %{
+            token: "Sim-123456",
+            team_id: "TXXXXXXXX",
+            api_app_id: "AXXXXXX",
+            event: %{
+              client_msg_id: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
+              type: "message",
+              text: "hi",
+              user: "UXXXXXXXX",
+              ts: "1559841924.001600",
+              channel: "asdf",
+              event_ts: "1559841924.001600",
+              channel_type: "channel"
+            },
+            type: "event_callback",
+            event_id: "EvXXXXXX",
+            event_time: 1_559_841_923,
+            authed_users: ["UXXXXXXXX", "UXXXXXXXX"]
+          }
+        }
+      )
+
+      events =
+        Event
+        |> Repo.all()
+
+      assert length(events) == 1
+      first_event = events |> Enum.at(0)
+      assert Map.fetch(first_event, :source_channel) == {:ok, "general"}
+    end
+  end
+
+  test "check channel name (nil) was added to events table", %{conn: conn} do
+    body = %{
+      "channel" => "asdf"
+    }
+
+    request_body = URI.encode_query(body)
+
+    headers = [
+      {"Accept", "application/json"},
+      {"Authorization", "Bearer asdf"},
+      {"Content-Type", "application/x-www-form-urlencoded"}
+    ]
+
+    with_mock HTTPoison,
+      post!: fn "https://slack.com/api/conversations.info",
+                request_body,
+                headers,
+                follow_redirect: true ->
+        %HTTPoison.Response{
+          body: "{
+            \"ok\": \"true\",
+            \"channel\": {
+
+            }
+        }"
+        }
+      end do
```

Comment:
Could you explain the syntax of this with_mock function to me, either here or in slack?

Reply:
the slack controller makes an api call to slack, and the with_mock function allows me to circumvent this actual call and replace it with this "mock" call.